librbd: quiesce/unquiesce API improvements

trociny opened the pull request:
<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

trociny commented:
@dillaman Thinking about primary mirror snapshot, I suppose we also want to provide a possibility to skip quiesce or ignore error for those. I could add `rbd_mirror_image_create_snapshot2` that would include the `flags` param. But the snapshots are also created on other events like demote/promote, which users might also want to control...

I think we should not care about secondary snapshots created by rbd-mirror as the quiesce callbacks are not supposed here.

dillaman commented:
> @dillaman Thinking about primary mirror snapshot, I suppose we also want to provide a possibility to skip quiesce or ignore error for those. I could add `rbd_mirror_image_create_snapshot2` that would include the `flags` param. But the snapshots are also created on other events like demote/promote, which users might also want to control...

Demote and promote, at least, should imply IO is stopped, so I wouldn't be worried about those cases.

> I think we should not care about secondary snapshots created by rbd-mirror as the quiesce callbacks are not supposed here.

Agreed.

dillaman commented on src/librbd/ImageWatcher.cc:
```diff
@@ -366,8 +366,10 @@ void ImageWatcher<I>::notify_quiesce(const AsyncRequestId &async_request_id,
                              << dendl;
 
   ceph_assert(attempts > 0);
+  auto notify_response = new watcher::NotifyResponse();
```
Nit: where is this freed?

dillaman replied:
... still open?

trociny replied:
note, I passed it to the `unique_ptr` in the lambda context below. I suppose it should do the magic?

dillaman replied:
Ack -- missed the generalized lambda capture

dillaman commented on src/librbd/api/Snapshot.cc:
```diff
@@ -309,6 +311,33 @@ int Snapshot<I>::exists(I *ictx, const cls::rbd::SnapshotNamespace& snap_namespa
+    internal_flags |= SNAP_CREATE_FLAG_SKIP_NOTIFY_QUIESCE;
+    flags &= ~RBD_SNAP_CREATE_SKIP_QUIESCE;
+  }
+  if (flags & RBD_SNAP_CREATE_IGNORE_QUIESCE_ERROR) {
```
Nit: `else if`? That will allow it to hit the invalid flag check below since it doesn't make sense to skip quiesce + ignore quiesce errors.

trociny commented:
> Demote and promote, at least, should imply IO is stopped, so I wouldn't be worried about those cases.

Just to make clear. 

Are you suggesting to provide `int rbd_mirror_image_create_snapshot2(uint32 flags, uint64_t snap_id)`, and when creating promote/demote snapshots  always use `SNAP_CREATE_FLAG_SKIP_NOTIFY_QUIESCE`?

dillaman commented:
> Are you suggesting to provide `int rbd_mirror_image_create_snapshot2(uint32 flags, uint64_t snap_id)`, and when creating promote/demote snapshots always use `SNAP_CREATE_FLAG_SKIP_NOTIFY_QUIESCE`?

For `promote`, you know the image is non-primary and therefore read-only -- so it can be hard-coded to `SNAP_CREATE_FLAG_SKIP_NOTIFY_QUIESCE`. For `demote`, we enable the read-only feature bit and then create the snapshot -- so same rationale I think. 

The mirror enable case might be another one where technically we should extend the API to add a quiesce flag?

dillaman commented:
... one more nit: `deep_copy/SnapshotCreateRequest` should probably hard-code the `SNAP_CREATE_FLAG_SKIP_NOTIFY_QUIESCE` flag

dillaman commented:
> ... one more nit: `deep_copy/SnapshotCreateRequest` should probably hard-code the `SNAP_CREATE_FLAG_SKIP_NOTIFY_QUIESCE` flag

... and `journal/Replay.cc`

trociny commented:
> The mirror enable case might be another one where technically we should extend the API to add a quiesce flag?

It looks so, though I am not very happy about this. It might look a bit awkward, especially if one tries to pass it when enabling journal mirroring.

trociny commented:
Also we could use `IGNORE_NOTIFY_QUIESCE_ERROR` when the old `mirror_image_create_snapshot` method (without flags) is called.

dillaman commented:
> May be just set `IGNORE_NOTIFY_QUIESCE_ERROR` for mirror enable?

... or just add a config `rbd_default_snapshot_quiesce_mode` to allow a blanket default used everywhere. I could actually see wanting to only error-out if mirror enable failed to quiesce since, unlike the mirror snapshot scheduler, it's something the admin would probably want to handle right away.

trociny commented:
I am not sure `rbd_default_snapshot_quiesce_mode` will help with mirror enable. It is going to apply for all cases, but I would like to have different behavior for mirror enable and user/mirror snapshot create requests. For mirror enable I would prefer if it ignored the error. For `mirror_image_create_snapshot`, `snap_create` -- if if failed on the error.

And taking that we are providing `mirror_image_create_snapshot2`, `snap_create2`, where the quiesce mode can be set, `rbd_default_snapshot_quiesce_mode` does not look so useful, though I am fine with adding it, if it is used by `mirror_image_create_snapshot`, `snap_create` only. I think I would not like if it was also used by `mirror_image_enabale`.

As for the mirror snapshot scheduler I thought it would be changed to use `mirror_image_create_snapshot2` and the flag is configured, e.g. just a "global" mgr module option, and it could be override by a schedule, i.e when a schedule is added it could also pass the "quiesce_mode" option:
```
rbd mirror snapshot schedule add rbd/test 1h --quiesce-mode skip
```
But I think just relaying on `rbd_default_snapshot_quiesce_mode` would work too -- it could be overridden per pool/namespace/image with `rbd config` commands.

trociny commented:
@dillaman update is pushed. I think I addressed all comments but implementing `rbd_default_snapshot_quiesce_mode`, which I am happy to add when we reach agreement. Also, I added quiesce options to `rbd snap create` and `rbd mirror image snapshot` commands.

trociny commented:
BTW, when updating python bindings, I just added optional `flags` parameter to `create_snap`
and `mirror_image_create_snapshot`, which I think does not break compatibility. On the other hand, for `remove_snap` we mimic the librbd API, i.e. added `remove_snap2`. This looks inconsistent. May be we add optional `flags` to `remove_snap` and "deprecate" `remove_snap2` if it is possible?

dillaman commented:
> For mirror enable I would prefer if it ignored the error. For mirror_image_create_snapshot, snap_create -- if if failed on the error.

I would think you'd want them consistent regardless. One less special case to explain.

> As for the mirror snapshot scheduler I thought it would be changed to use mirror_image_create_snapshot2 and the flag is configured

Perfect -- even better.

> But I think just relaying on `rbd_default_snapshot_quiesce_mode` would work too -- it could be overridden per pool/namespace/image with `rbd config` commands.

Honestly, that's why I was more interested in the config value. i.e. perhaps the majority of users would just want to set "skip" to avoid the extra handshaking required for snapshot creation if they don't have any hook users.

dillaman commented:
> BTW, when updating python bindings, I just added optional `flags` parameter to `create_snap`
> and `mirror_image_create_snapshot`, which I think does not break compatibility. 

Yes, that's fine.

> On the other hand, for `remove_snap` we mimic the librbd API, i.e. added `remove_snap2`. This looks inconsistent. May be we add optional `flags` to `remove_snap` and "deprecate" `remove_snap2` if it is possible?

Yes, that sounds like a good plan to me.

trociny commented:
> I would think you'd want them consistent regardless. One less special case to explain.

Ok. I will add `rbd_default_snapshot_quiesce_mode` then.

> Honestly, that's why I was more interested in the config value.

So, after I add the `rbd_default_snapshot_quiesce_mode` config override, are we still interested in making the mirror snapshot scheduler to have it configurable too (per global or a schedule level) or we can just rely on the config override?

dillaman commented:
> So, after I add the `rbd_default_snapshot_quiesce_mode` config override, are we still interested in making the mirror snapshot scheduler to have it configurable too (per global or a schedule level) or we can just rely on the config override?

If you don't mind adding it, I think it provides a nice analogue to your rbd CLI changes.

trociny commented:
added `rbd_default_snapshot_quiesce_mode`

dillaman commented on src/common/options.cc:
```diff
@@ -7570,6 +7570,11 @@ static std::vector<Option> get_rbd_options() {
 
+    Option("rbd_default_snapshot_quiesce_mode", Option::TYPE_STR, Option::LEVEL_ADVANCED)
+    .set_default("obey")
+    .set_enum_allowed({"obey", "try", "skip"})
```
Nit: I would use `require(d)` instead of `obey` and perhaps `ignore-error` instead of `try` (or alternatively expand the description)

dillaman commented on src/tools/rbd_mirror/image_sync/SyncPointCreateRequest.cc:
```diff
@@ -113,7 +113,8 @@ void SyncPointCreateRequest<I>::send_create_snap() {
       this);
   m_remote_image_ctx->operations->snap_create(
-    cls::rbd::UserSnapshotNamespace(), sync_point.snap_name.c_str(), ctx);
+    cls::rbd::UserSnapshotNamespace(), sync_point.snap_name.c_str(), 0,
```
Since this is only used for journaling mirror-mode, we can skip the quiesce request.

dillaman commented on src/tools/rbd/ArgumentTypes.cc:
```diff
@@ -321,6 +321,13 @@ void add_flatten_option(boost::program_options::options_description *opt) {
 
+void add_snap_create_options(po::options_description *opt) {
+  opt->add_options()
+    (SKIP_QUIESCE.c_str(), po::bool_switch(), "do not run quiesce callbacks")
```
Nit: perhaps
```suggestion
    (SKIP_QUIESCE.c_str(), po::bool_switch(), "do not run quiesce hooks")
```

trociny commented:
@dillaman All other comments are addressed. The `make check` failed on `TestMigration.StressLive`. I suppose it a non-related issue, reported earlier [1].

[1] https://tracker.ceph.com/issues/45694

dillaman commented:
Looks like the CLI tests failed (and that random most likely unrelated `TestMigration.StressLive` failure)

trociny commented:
Ah, I saw `TestMigration.StressLive` failure and missed the CLI one. It should be fixed now. Thanks.

trociny commented:
@dillaman Pushed an additional commit that fixes a race I observed in TestLibRBD.QuiesceWatchError when running on one cpu.

dillaman commented:
Might be related? Getting `-ESTALE` errors propagated back to the user. 

http://qa-proxy.ceph.com/teuthology/jdillaman-2020-06-09_13:48:13-rbd-wip-jd-testing-distro-basic-smithi/5131926/teuthology.log

trociny commented:
@dillaman Thanks! I added a commit that just makes the watcher return 0 for a duplicate quiesce. It looks like no much use in returning `-ESTALE` here, which would need to be filtered out in `ImageWatcher::notify_quiesce` anyway.

dillaman commented:
Looks like the CLI integration test needs to be updated: http://qa-proxy.ceph.com/teuthology/jdillaman-2020-06-10_20:56:14-rbd-wip-jd-testing-distro-basic-smithi/5135019/teuthology.log

And a small memory leak: http://qa-proxy.ceph.com/teuthology/jdillaman-2020-06-10_20:56:14-rbd-wip-jd-testing-distro-basic-smithi/5135136/teuthology.log

trociny commented:
@dillaman Thanks. I think I fixed the issues you found with two additional commits I pushed a couple of days ago. But when I was testing it on teuthology I found yet another issue with a duplicate sent after quiesce failed. I pushed the fix for it in the last commit.

Here are more details for this failure. The example is here [1]. See the extract from the logs below. In the failed test the quiesce returned error, but due to "injected socket failure" the watcher resent the notification and it was processed as a new one instead of being ignored as a duplicate, because after quiesce error I did not track the request anymore. 

I fixed it by delaying the request unregistration. Not sure if you like it, thus added as a separate commit. I am tetsing it on teuthology right now.

```
2020-06-15T07:50:35.213 INFO:tasks.workunit.client.0.smithi205.stdout:[ RUN      ] TestLibRBD.QuiesceWatchError
2020-06-15T07:50:37.216 INFO:tasks.workunit.client.0.smithi205.stdout:using new format!
2020-06-15T07:50:37.261 INFO:tasks.workunit.client.0.smithi205.stdout:/home/jenkins-build/build/workspace/ceph-dev-new-build/ARCH/x86_64/AVAILABLE_ARCH/x86_64/AVAILABLE_DIST/centos8/DIST/centos8/MACHINE_SIZE/gigantic/release/16.0.0-2340-gec00b5d/rpm/el8/BUILD/ceph-16.0.0-2340-gec00b5d/src/test/librbd/test_librbd.cc:8365: Failure
2020-06-15T07:50:37.261 INFO:tasks.workunit.client.0.smithi205.stdout:Expected equality of these values:
2020-06-15T07:50:37.262 INFO:tasks.workunit.client.0.smithi205.stdout:  1U
2020-06-15T07:50:37.262 INFO:tasks.workunit.client.0.smithi205.stdout:    Which is: 1
2020-06-15T07:50:37.262 INFO:tasks.workunit.client.0.smithi205.stdout:  watcher1.quiesce_count
2020-06-15T07:50:37.262 INFO:tasks.workunit.client.0.smithi205.stdout:    Which is: 2
2020-06-15T07:50:37.265 INFO:tasks.workunit.client.0.smithi205.stdout:[  FAILED  ] TestLibRBD.QuiesceWatchError (2052 ms)

2020-06-15T07:50:37.255+0000 7f3ae24c8b00  5 librbd::SnapshotCreateRequest: 0x7f3a4c01ae70 send_notify_quiesce
2020-06-15T07:50:37.255+0000 7f3ae24c8b00 10 librbd::ImageWatcher: 0x7f3a78006660 notify_quiesce: request_id=0x7f3a4c01af90
2020-06-15T07:50:37.255+0000 7f3ae24c8b00 10 librbd::ImageWatcher: 0x7f3a78006660 notify_quiesce: async_request_id=[4781,139889232356272,2] attempts=10
2020-06-15T07:50:37.255+0000 7f3ae24c8b00 20 librbd::watcher::Notifier: 0x7f3a780066e0 notify: pending=1
2020-06-15T07:50:37.256+0000 7f3a927fc700  5 librbd::Watcher: 0x7f3a78006660 notifications_blocked: blocked=0
2020-06-15T07:50:37.256+0000 7f3a927fc700 10 librbd::Watcher::C_NotifyAck 0x560dddbf4bd0 C_NotifyAck: id=1666447310901, handle=139889232356272
2020-06-15T07:50:37.256+0000 7f3a927fc700 10 librbd::ImageWatcher: 0x7f3a78006660 quiesce request: [4781,139889232356272,2]
2020-06-15T07:50:37.256+0000 7f3a927fc700 20 librbd::ImageState: 0x7f3a80003fa0 notify_quiesce
2020-06-15T07:50:37.256+0000 7f3a927fc700 20 librbd::ImageState: 0x560dddd50440 QuiesceWatchers::notify event: 0
2020-06-15T07:50:37.256+0000 7f3a927fc700  5 librbd::Watcher: 0x7f3a78006e20 notifications_blocked: blocked=0
2020-06-15T07:50:37.256+0000 7f3a927fc700 10 librbd::Watcher::C_NotifyAck 0x7f3a78005770 C_NotifyAck: id=1666447310901, handle=139889232357632
2020-06-15T07:50:37.256+0000 7f3a927fc700 10 librbd::ImageWatcher: 0x7f3a78006e20 quiesce request: [4781,139889232356272,2]
2020-06-15T07:50:37.256+0000 7f3a927fc700 20 librbd::ImageState: 0x7f3a5402fd70 notify_quiesce
2020-06-15T07:50:37.256+0000 7f3a7cff9700 20 librbd::ImageState: 0x560dddd50440 QuiesceWatchers::operator(): handle=0, event_type=0
2020-06-15T07:50:37.256+0000 7f3a927fc700 20 librbd::ImageState: 0x7f3a88052c10 QuiesceWatchers::notify event: 0
2020-06-15T07:50:37.256+0000 7f3a7cff9700 10 librbd::ImageWatcher: 0x7f3a7800aef0 C_ResponseMessage: r=-22
2020-06-15T07:50:37.256+0000 7f3a7cff9700 10 librbd::Watcher::C_NotifyAck 0x560dddbf4bd0 finish: r=0
2020-06-15T07:50:37.256+0000 7f3a7cff9700 20 librbd::ImageState: 0x7f3a88052c10 QuiesceWatchers::operator(): handle=0, event_type=0
2020-06-15T07:50:37.256+0000 7f3a7cff9700 10 librbd::ImageWatcher: 0x7f3a780147c0 C_ResponseMessage: r=0
2020-06-15T07:50:37.256+0000 7f3a7cff9700 10 librbd::Watcher::C_NotifyAck 0x7f3a78005770 finish: r=0
2020-06-15T07:50:37.256+0000 7f3a9cff9700  0 -- 172.21.15.205:0/2227350295 >> [v2:172.21.15.66:6816/30643,v1:172.21.15.66:6819/30643] conn(0x7f3aa401a3a0 msgr2=0x7f3aa404ffd0 crc :-1 s=STATE_CONNECTION_ESTABLISHED l=1).read_until injecting socket failure
2020-06-15T07:50:37.258+0000 7f3a91ffb700  5 librbd::Watcher: 0x7f3a78006660 notifications_blocked: blocked=0
2020-06-15T07:50:37.258+0000 7f3a91ffb700 10 librbd::Watcher::C_NotifyAck 0x560dddbf74b0 C_NotifyAck: id=1666447310902, handle=139889232356272
2020-06-15T07:50:37.258+0000 7f3a91ffb700 10 librbd::ImageWatcher: 0x7f3a78006660 quiesce request: [4781,139889232356272,2]
2020-06-15T07:50:37.258+0000 7f3a91ffb700 20 librbd::ImageState: 0x7f3a80003fa0 notify_quiesce
2020-06-15T07:50:37.258+0000 7f3a91ffb700 20 librbd::ImageState: 0x560dddd50440 QuiesceWatchers::notify event: 0
2020-06-15T07:50:37.258+0000 7f3a91ffb700  5 librbd::Watcher: 0x7f3a78006e20 notifications_blocked: blocked=0
2020-06-15T07:50:37.258+0000 7f3a91ffb700 10 librbd::Watcher::C_NotifyAck 0x7f3a60002b20 C_NotifyAck: id=1666447310902, handle=139889232357632
2020-06-15T07:50:37.258+0000 7f3a91ffb700 10 librbd::ImageWatcher: 0x7f3a78006e20 duplicate quiesce request: [4781,139889232356272,2]
2020-06-15T07:50:37.258+0000 7f3a7dffb700 10 librbd::ImageWatcher: 0x7f3a6004f200 C_ResponseMessage: r=0
2020-06-15T07:50:37.258+0000 7f3a7cff9700 20 librbd::ImageState: 0x560dddd50440 QuiesceWatchers::operator(): handle=0, event_type=0
2020-06-15T07:50:37.258+0000 7f3a7dffb700 10 librbd::Watcher::C_NotifyAck 0x7f3a60002b20 finish: r=0
2020-06-15T07:50:37.258+0000 7f3a7cff9700 10 librbd::ImageWatcher: 0x7f3a6004dfe0 C_ResponseMessage: r=-22
2020-06-15T07:50:37.258+0000 7f3a7cff9700 10 librbd::Watcher::C_NotifyAck 0x560dddbf74b0 finish: r=0
2020-06-15T07:50:37.259+0000 7f3a927fc700 20 librbd::watcher::Notifier: 0x7f3a780066e0 handle_notify: r=0
2020-06-15T07:50:37.259+0000 7f3a927fc700 20 librbd::watcher::Notifier: 0x7f3a780066e0 handle_notify: pending=0
2020-06-15T07:50:37.259+0000 7f3a7cff9700 -1 librbd::ImageWatcher: 0x7f3a78006660 failed to notify quiesce: (22) Invalid argument
2020-06-15T07:50:37.259+0000 7f3a7cff9700  5 librbd::SnapshotCreateRequest: 0x7f3a4c01ae70 handle_notify_quiesce: r=-22
```

[1] http://qa-proxy.ceph.com/teuthology/trociny-2020-06-15_07:26:08-rbd-wip-mgolub-testing-distro-basic-smithi/5149860/teuthology.log

trociny commented:
> I fixed it by delaying the request unregistration. Not sure if you like it, thus added as a separate commit. I am tetsing it on teuthology right now.

Ah, it still does not work as expected [1]. Now if a duplicate request for an "error" quiesce comes, it is responded without the error, because the error code is not stored anywhere, and I don't see where I could keep it without adding a special structure like `std::map<watch_notify::AsyncRequestId, int> m_quiesce_pending` so it could store the failure code there. And if adding this (and use instead of `m_async_pending`) then may be also add `std::set<watch_notify::AsyncRequestId> m_quiesce_requests` and use instead of reusing `std::map<watch_notify::AsyncRequestId, AsyncRequest> m_async_requests`?

@dillaman what do you think?

[1] http://qa-proxy.ceph.com/teuthology/trociny-2020-06-15_12:40:08-rbd-wip-mgolub-testing-distro-basic-smithi/5150286/teuthology.log

dillaman commented:
In the real-world, this case seems like it would be equivalent to two snapshot create request where the first one fails and the second one succeeds. Perhaps the test can be tweaked to just return the error code repeatedly.

trociny commented:
I don't think it is just the test problem. The quiesce hook returned an error, which means that the snapshot create should fail (if the user has not specified some other behavior), and due to an intermittent networking issue the error code is just lost and the shanpshot create proceeds.

We could do a different way. Before I added this last patch, the behavior on receiving a duplicate for a failed quiesce had been to try to execute the quiesce again (which would most likely fail again). We could just drop this last patch and tweak the test not to fail in such a case. Not sure it is a good idea though in real life to run the quiesce hook once more time when it is failing...

dillaman commented:
> We could do a different way. Before I added this last patch, the behavior on receiving a duplicate for a failed quiesce had been to try to execute the quiesce again (which would most likely fail again). We could just drop this last patch and tweak the test not to fail in such a case. Not sure it is a good idea though in real life to run the quiesce hook once more time when it is failing...

I agree that in the perfect world we shouldn't run `quiesce` again, but in this case where there was a network hiccup, I don't think it's worth much effort to try to work around it and keep the last X seconds of return code state.

trociny commented:
Ok. I removed the last patch and added the one that tweak the test instead. I will test it on teuthology.

trociny commented:
@dillaman I re-ran the tests that previously failed on teuthology and they work now. Also ran a more generic subset of test [1] just to make sure I did not break anything with my last commits and it looks like there are no related failures.

[1] http://pulpito.ceph.com/trociny-2020-06-18_09:19:50-rbd-wip-mgolub-testing-distro-basic-smithi/

dillaman commented:
jenkins test make check
